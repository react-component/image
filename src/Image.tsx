import * as React from 'react';
import { useState } from 'react';
import cn from 'classnames';
import { getOffset } from 'rc-util/lib/Dom/css';
import Preview from './Preview';

export interface ImageProps
  extends Omit<React.ImgHTMLAttributes<HTMLImageElement>, 'placeholder' | 'onClick'> {
  // Original
  src?: string;

  prefixCls?: string;
  previewPrefixCls?: string;
  placeholder?: React.ReactNode;
  fallback?: string;
  preview?: boolean;
  onPreviewClose?: (e: React.SyntheticEvent<HTMLDivElement | HTMLLIElement>) => void;
  onClick?: (e: React.MouseEvent<HTMLDivElement>) => void;
}

type ImageStatus = 'normal' | 'error' | 'loading';

const ImageInternal: React.FC<ImageProps> = ({
  src,
  alt,
  onPreviewClose: onInitialPreviewClose,
  prefixCls = 'rc-image',
  previewPrefixCls = `${prefixCls}-preview`,
  placeholder,
  fallback,
  width,
  height,
  style,
  preview = true,
  className: originalClassName,
  onClick,

  // Img
  crossOrigin,
  decoding,
  loading,
  referrerPolicy,
  sizes,
  srcSet,
  useMap,
  ...otherProps
}) => {
  const isCustomPlaceholder = placeholder && placeholder !== true;
  const [isShowPreview, setShowPreview] = useState(false);
  const [status, setStatus] = useState<ImageStatus>(isCustomPlaceholder ? 'loading' : 'normal');
  const [mousePosition, setMousePosition] = useState<null | { x: number; y: number }>(null);
  const isError = status === 'error';

  const onLoad = () => {
    setStatus('normal');
  };

  const onError = () => {
    setStatus('error');
  };

  const onPreview: React.MouseEventHandler<HTMLDivElement> = e => {
    const { left, top } = getOffset(e.target);

    setShowPreview(true);
    setMousePosition({
      x: left,
      y: top,
    });

    if (onClick) onClick(e);
  };

  const onPreviewClose = (e: React.SyntheticEvent<HTMLDivElement>) => {
    e.stopPropagation();
    setShowPreview(false);
    setMousePosition(null);

    if (onInitialPreviewClose) onInitialPreviewClose(e);
  };

  React.useEffect(() => {
    if (isCustomPlaceholder) {
      setStatus('loading');
    }
  }, [src]);

  const className = cn(prefixCls, originalClassName, {
    [`${prefixCls}-error`]: isError,
  });

  const mergedSrc = isError && fallback ? fallback : src;
  const imgCommonProps = {
    crossOrigin,
    decoding,
    loading,
    referrerPolicy,
    sizes,
    srcSet,
    useMap,
    alt,
    className: cn(`${prefixCls}-img`, {
      [`${prefixCls}-img-placeholder`]: placeholder === true,
    }),
  };

  return (
    <>
      <div
        {...otherProps}
        className={className}
        onClick={preview && !isError ? onPreview : onClick}
        style={{
          ...style,
          width,
          height,
        }}
      >
        {isError && fallback ? (
          <img {...imgCommonProps} src={fallback} />
        ) : (
          <img {...imgCommonProps} onLoad={onLoad} onError={onError} src={src} />
        )}

        {status === 'loading' && (
          <div aria-hidden="true" className={`${prefixCls}-placeholder`}>
            {placeholder}
          </div>
        )}
      </div>
      {preview && !isError && (
        <Preview
          aria-hidden={!isShowPreview}
          visible={isShowPreview}
          prefixCls={previewPrefixCls}
          onClose={onPreviewClose}
          mousePosition={mousePosition}
          src={mergedSrc}
          alt={alt}
        />
      )}
    </>
  );
};

ImageInternal.displayName = 'Image';

export default ImageInternal;
