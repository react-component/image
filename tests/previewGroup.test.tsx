import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import KeyCode from 'rc-util/lib/KeyCode';
import Image from '../src';

describe('PreviewGroup', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('Mount and UnMount', () => {
    const wrapper = mount(
      <Image.PreviewGroup>
        <Image src="src1" />
        <Image src="src2" />
      </Image.PreviewGroup>,
    );

    act(() => {
      wrapper.find('.rc-image').at(0).simulate('click');
      jest.runAllTimers();
      wrapper.update();
    });
    expect(wrapper.find('.rc-image-preview').get(0)).toBeTruthy();

    const previewProgressElement = wrapper.find(
      '.rc-image-preview .rc-image-preview-operations-progress',
    );

    expect(previewProgressElement).toBeTruthy();
    expect(previewProgressElement.text()).toEqual('1 / 2');

    expect(() => {
      wrapper.unmount();
    }).not.toThrow();
  });

  it('Disable preview', () => {
    const wrapper = mount(
      <Image.PreviewGroup>
        <Image src="src1" preview={false} />
      </Image.PreviewGroup>,
    );

    act(() => {
      wrapper.find('.rc-image').at(0).simulate('click');
      jest.runAllTimers();
      wrapper.update();
    });

    expect(wrapper.find('.rc-image-preview').get(0)).toBeFalsy();
  });

  it('Preview with Custom Preview Property', () => {
    const wrapper = mount(
      <Image.PreviewGroup
        preview={{
          countRender: (current, total) => `current:${current} / total:${total}`,
        }}
      >
        <Image src="src1" />
        <Image src="src2" />
        <Image src="src2" />
      </Image.PreviewGroup>,
    );

    act(() => {
      wrapper.find('.rc-image').at(0).simulate('click');
      jest.runAllTimers();
      wrapper.update();
    });

    const previewProgressElement = wrapper.find(
      '.rc-image-preview .rc-image-preview-operations-progress',
    );

    expect(previewProgressElement).toBeTruthy();
    expect(previewProgressElement.text()).toEqual('current:1 / total:3');
  });

  it('Switch', () => {
    const previewProgressElementPath = '.rc-image-preview .rc-image-preview-operations-progress';
    const wrapper = mount(
      <Image.PreviewGroup>
        <Image src="src1" />
        <Image src="src2" preview={false} />
        <Image src="src3" />
      </Image.PreviewGroup>,
    );

    act(() => {
      wrapper.find('.rc-image').at(0).simulate('click');
      jest.runAllTimers();
      wrapper.update();
    });

    expect(
      wrapper.find('.rc-image-preview .rc-image-preview-switch-left-disabled').get(0),
    ).toBeTruthy();
    expect(wrapper.find(previewProgressElementPath).text()).toEqual('1 / 2');

    act(() => {
      wrapper.find('.rc-image-preview .rc-image-preview-switch-right').simulate('click');
      jest.runAllTimers();
      wrapper.update();
    });

    expect(
      wrapper.find('.rc-image-preview .rc-image-preview-switch-right-disabled').get(0),
    ).toBeTruthy();
    expect(wrapper.find(previewProgressElementPath).text()).toEqual('2 / 2');

    act(() => {
      wrapper.find('.rc-image-preview .rc-image-preview-switch-left').simulate('click');
      jest.runAllTimers();
      wrapper.update();
    });

    expect(
      wrapper.find('.rc-image-preview .rc-image-preview-switch-left-disabled').get(0),
    ).toBeTruthy();

    act(() => {
      const event = new KeyboardEvent('keydown', { keyCode: KeyCode.RIGHT });
      global.dispatchEvent(event);
      jest.runAllTimers();
      wrapper.update();
    });

    expect(
      wrapper.find('.rc-image-preview .rc-image-preview-switch-right-disabled').get(0),
    ).toBeTruthy();

    act(() => {
      const event = new KeyboardEvent('keydown', { keyCode: KeyCode.LEFT });
      global.dispatchEvent(event);
      jest.runAllTimers();
      wrapper.update();
    });

    expect(
      wrapper.find('.rc-image-preview .rc-image-preview-switch-left-disabled').get(0),
    ).toBeTruthy();
  });

  it('With Controlled', () => {
    const wrapper = mount(
      <Image.PreviewGroup preview={{ visible: true }}>
        <Image src="src1" />
      </Image.PreviewGroup>,
    );

    expect(wrapper.find('.rc-image-preview').get(0)).toBeTruthy();

    act(() => {
      wrapper.setProps({ preview: { visible: false } });
      jest.runAllTimers();
      wrapper.update();
    });

    expect(wrapper.find('.rc-image-preview').at(0).render()).toMatchSnapshot();
  });
});
