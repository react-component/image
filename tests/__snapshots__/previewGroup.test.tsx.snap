// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewGroup With Controlled 1`] = `
<div
  class="rc-image-preview"
  role="document"
>
  <div
    aria-hidden="true"
    style="width: 0px; height: 0px; overflow: hidden; outline: none;"
    tabindex="0"
  />
  <div
    class="rc-image-preview-content"
  >
    <div
      class="rc-image-preview-body"
    >
      <ul
        class="rc-image-preview-operations"
      >
        <li
          class="rc-image-preview-operations-operation"
        />
        <li
          class="rc-image-preview-operations-operation"
        />
        <li
          class="rc-image-preview-operations-operation rc-image-preview-operations-operation-disabled"
        />
        <li
          class="rc-image-preview-operations-operation"
        />
        <li
          class="rc-image-preview-operations-operation"
        />
      </ul>
      <div
        class="rc-image-preview-img-wrapper"
        style="transform: translate3d(0px, 0px, 0);"
      >
        <img
          class="rc-image-preview-img"
          src="src1"
          style="transform: scale3d(1, 1, 1) rotate(0deg);"
        />
      </div>
    </div>
  </div>
  <div
    aria-hidden="true"
    style="width: 0px; height: 0px; overflow: hidden; outline: none;"
    tabindex="0"
  />
</div>
`;
